### ISSUE TITLE
Support additional module-info.java for tests, if that ends up being de-facto standard

### ISSUE THREAD

**Issue by hohwille** [eclipse-jdt/eclipse.jdt.core#1465]
# Problem
If a module requires additional modules only for its test cases, the de-facto standard is to define an additional `module-info.java` file in `src/test/java` that should be declared open and has a test specific identifier and requires the main module from `src/main/java/module-info.java`.
Here is an official example from Maven for this use-case: https://maven.apache.org/surefire/maven-surefire-plugin/examples/jpms.html

However, Eclipse does not support that a single project may have more than one `module-info.java`.
Therefore you get the following error:
```
Build path contains duplicate entry: 'module-info.java' for project '«project-name»'
```

# Why is this important?
The JPMS is very strict. As soon as I define modules, I need to declare any required module or I can not even import an according class.
So in a library project, I might want to have a very minimal required module dependencies (e.g. none at all). However, for implementing my JUnit tests, I might need some logging (`java.util.logging` or `org.slf4j`). I do not want to be forced to require that logging in my module just to be able to use it only in my tests. Otherwise, I can not even add the import to the Logger in my JUnit test class. The logging is just a famous example - this also applies to whatever else you might need additionally for your tests.
With JPMS the entire concept of test scoped dependencies becomes more or less pointless with Eclipse due to this lacking feature.

# Impact
Developers of JPMS modules are forced to use a different IDE than Eclipse.
A workaround may be to create an extra projects only for testing. However this has the following drawbacks:
* It does not fit with common standards and confuses developers trying to understand the project
* It causes lots of clutter and maintenance overhead
* In an extra project I can not access the internals of the actual project, I want to test, so I am limited to black-box-testing. However, I commonly want to write JUnits as white-box tests.

# Related issues
* This issue is a replacement for [bugzilla 559601](https://bugs.eclipse.org/bugs/show_bug.cgi?id=559601).
* It is slightly related to #617

# Sources
* https://github.com/eclipse-jdt/eclipse.jdt.core/blob/aea7697979eae6edc916094f25a716d529c40b68/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/util/messages.properties#L155
* https://github.com/eclipse-jdt/eclipse.jdt.core/blob/aea7697979eae6edc916094f25a716d529c40b68/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaProject.java#L3996

**Comment by stephan-herrmann** [comment]
> the de-facto standard is to define an additional module-info.java file in src/test/java that should be declared open and has a test specific identifier and requires the main module from src/main/java/module-info.java.

It's a pity, that no real standard has emerged, and so we are left with various concepts that cannot easily be reconciled.

If I understand the maven approach correctly, then src/main/java and src/test/java should be seen as two regular modules (JMPS) that just happen to share a common pom (which in itself may have a test-scoped slice that doesn't apply to src/main/java).

This differs from another concept that has been proposed: to consider src/test/java as an overlay, which is merged with src/main/java for compilation and execution of tests and ignored otherwise. 

I think it's fair to call the first approach **black-box** tests whereas the second approach (merging) support **white-box** testing (of private innards of a module).
 
Traditionally, in Eclipse the **black-box** approach most closely corresponds to two separate projects. The key point being: in Eclipse a classpath is defined exactly once per project. Eclipse does not genuinely define any conditional logic in `.classpath`. There are two backdoors, though:

- JDT leaves it to the implementation of "Classpath Containers", to resolve into an arbitrary set of jars or classpath folders, at the full deliberation of that implementation.
- JDT hardcodes specific semantics for a source folder marked as "test", see https://eclipse.dev/eclipse/news/4.8/jdt.php#jdt-test-sources

Let's briefly dwell on the latter feature. References (in addition to the N&N entry above):
* see comments starting at https://bugs.eclipse.org/bugs/show_bug.cgi?id=224708#c45
* in JDT sources search for references to o.e.j.i.c.CompilationGroup.TEST
* JPMS specific details are in 
   * https://bugs.eclipse.org/bugs/show_bug.cgi?id=520713 
   * https://bugs.eclipse.org/bugs/show_bug.cgi?id=527593

For a first approximation of the semantics this is what @brychcy wrote:

> For test code I implemented the following in my patch:
>  * production sources see the workspace as if test sources and test dependencies don't exist
>  * test sources see the workspace as if the whole mechanism doesn't exist -
>  (actually it sees the production sources in the same project only as binaries instead of sources, but this shouldn't make a difference)

And then any dependency can be marked as visible only for the "test" scope. Tada!

With respect to JPMS this feature implements the suggestions from http://openjdk.java.net/jeps/261 (which may be seen as authoritive, no?) as to use options like `--add-reads` and `--patch-module` to integrate test code with module main code.

As you can see, the internal semantics of marking folders and dependencies as "test" or not test, is already quite sophisticated, but it seems to solve "the other problem": enabling white box testing while avoiding the need of a secondary `module-info.java`.

Back to the request to support a two-module black-box approach, the above seems to imply, that we first need to *disable* parts of the existing solution, viz. we need a mode *without* any privileges for the test folder. OTOH, we need to retain the ability to compile in two separate invocations for main vs test.

So **in the implementation** we'd need to extend `enum CompilationGroup { MAIN, TEST }` with a third option like `TEST_MODULE`. Then every code location asking for `TEST` needs to differentiate what needs to be done for the new mode. If requirements at this level can be specified in terms of a compiler command line then tweaking this part accordingly sounds manageable.

**In the UI** I would prefer not to surface all three modes, so as a quick idea, we *could* distinguish the two test modes based on existence of `src/test/java/module-info.java` (not based on the name of the source folder, but based on the classpath attribute "test").

Finally, there is the **API** perspective, where JDT promises that each modular `IJavaProject` has exactly one `IModuleDescription`. It looks like the test module simply can't be accessed via this interface. This will make it very hard to implement any additional support for the test module beyond just compilation. Inspecting the incoming call hierarchy of `org.eclipse.jdt.core.IJavaProject.getModuleDescription()` will give an impression of the impact.

 I should also mention, that JDT is build-tool agnostic, i.e., we can only provide very generic capabilities, so that m2e can setup a project's configuration according to maven rules. For m2e the main concept for communicating with JDT is through classpath entries, where things like "test" are communicated via classpath attributes on those entries. More specifically, m2e can
* create or update `.classpath` to affect the general setup (like marking every `src/test/java` as "test)
* resolve dependencies from pom.xml to entries of the **Maven Dependencies** container, again possibly adorned with classpath attributes, per resolved entry i.e.,

It needs to be determined here, if JDT needs to understand additional classpath attributes to serve m2e right.

_On a personal note: I will continue to discuss this and give hints into the implementation if desired, but I will not actively drive this development, in part because I feel that supporting just another project layout for black box testing (which can already be done by individual Eclipse projects) is not really a huge advance. I think white box testing in the presence of modules is the thing calling for _really_ new concepts, like merging `module-info.java` from two sources or such._

**Comment by stephan-herrmann** [comment]
> In an extra project I can not access the internals of the actual project, I want to test, so I am limited to black-box-testing. However, I commonly want to write JUnits as white-box tests.

Rereading that sentence from the issue description I wonder if I may have misunderstood anything in how maven supports testing vis-a-vis JPMS: In my book defining a separate module for tests is black-box testing (see above), since a regular (test-)module does not have any privilege over other modules. In that sense the JPMS module is stronger than the colocation in the same "project", because "project" is not a concept known to Java/JPMS.

Is white-box testing of modular code supported by maven?

**Comment by SentryMan** [comment]
Surefire 3 lets you disable the module path for easy white-box testing, which is what I end up doing most of the time. It is annoying that eclipse doesn't respect it though.

**Comment by stephan-herrmann** [comment]
> Surefire 3 lets you disable the module path for easy white-box testing, which is what I end up doing most of the time. It is annoying that eclipse doesn't respect it though.

This is the opposite request from this current issue (which is about respecting two `module-info.java` for what to me looks like black box testing).

@SentryMan Please file a separate issue for your use case. I believe "the spirit of" your use case is already supported in the IDE, so we'd need to see the exact configuration of a sample Eclipse project, to figure out the differences between a working configuration for white box testing vs. what you have in mind.

**Comment by hohwille** [comment]
> > Surefire 3 lets you disable the module path for easy white-box testing, which is what I end up doing most of the time. It is annoying that eclipse doesn't respect it though.
> 
> This is the opposite request from this current issue (which is about respecting two `module-info.java` for what to me looks like black box testing).
> 
> @SentryMan Please file a separate issue for your use case. I believe "the spirit of" your use case is already supported in the IDE, so we'd need to see the exact configuration of a sample Eclipse project, to figure out the differences between a working configuration for white box testing vs. what you have in mind.

I can confirm this. Regular white-box testing works fine. But assuming, I need to add a `requires` or `provides` or whatever to my `module-info.java` that I only want to have for testing, I am left in the rain. A workaround is add a second project/module but then I also have to be careful that my build is not accidentally deploying this test artifact as a release. Also the workaround is causing quite some overhead. But this is what I currently do to make it work.
The status quo is quite confusing for developers if maven documents some approach that is however not supported by IDEs what makes the entire approach kind of useless. IMHO the suggested approach sounds totally obvious and straight forward. It might be that the internal design of eclipse is not ready for this so huge refactoring effort would be required. I still see tons of other things in Eclipse that are broken for decades that may be of higher importance then...

**Comment by hohwille** [comment]
> In my book defining a separate module for tests is black-box testing (see above), since a regular (test-)module does not have any privilege over other modules. In that sense the JPMS module is stronger than the colocation in the same "project", because "project" is not a concept known to Java/JPMS.

I am unsure how maven behaves here exactly but my expectation would be that in both cases you are doing white-box testing and only if you define an additional `module-info.java` in `src/test/java` as well, then during the tests this will "override" the one from `src/main/java`.

**Comment by danthe1st** [comment]
> Perhaps there's a lack of enthusiasm because all that work would enable eclipse to do something which is already possible (define a module for black box testing) only in a different layout of projects / folders.

To be honest, I think that most (at least non-Eclipse/OSGi-related which aren't using JPMS anyways AFAIK) Java projects would prefer using the layout  with both sources and tests in the same project. I think there are many people who would like to use JPMS but wouldn't be able to do that in Eclipse due to this issue.
While the approach with `<useModulePath>false</useModulePath>` in the `<configuration>` of the `maven-surefire-plugin` seems to work (at least for me) with Eclipse, I think this option isn't really known.

Even though many projects still don't use `module-info.java` files, this seems to be requested more and more.

Also, if a project is using that approach, that would pretty much lock out Eclipse users from that project.

Given these points, I think many people would appreciate this to be implemented.

**Comment by stephan-herrmann** [comment]
For all those interested in simply hosting two regular JPMS modules in src/main/java and src/test/java with no further tweaks, please have a look at this simple example:
[MainAndTest-layout.zip](https://github.com/user-attachments/files/17085373/MainAndTest-layout.zip)

It does the simplest thing that could possibly work: create a nested project located in src/test.
Import the projects in 2 steps:
* from the archive import only project "Main"
* next **Import** > **Existing Project**, pointing to Main/src/test

I left one detail as an exercise to the reader: link the output folder src/test/target/classes to target/test-classes (I believe linked resources should help here, or an actual link in the filesystem, whatever works better).

I think m2e should be able to conveniently configure all this behind the scenes, i.e., it seems JDT is already able to support the requirements, no? 

In the Project Explorer you may want to select **Projects Presentation** > **Hierarchical**.
![image](https://github.com/user-attachments/assets/787b41cb-933d-45be-b10f-fe97b5cf3939)

**Comment by danthe1st** [comment]
> I think m2e should be able to conveniently configure all this behind the scenes, i.e., it seems JDT is already able to support the requirements, no?

Do you mean with a distinct `pom.xml` in `src/test` with a bit of extra configuration?

**Comment by stephan-herrmann** [comment]
> > I think m2e should be able to conveniently configure all this behind the scenes, i.e., it seems JDT is already able to support the requirements, no?
> 
> Do you mean with a distinct `pom.xml` in `src/test` with a bit of configuration in it? I think importing as a non-Maven project (and importing as a Maven project requires a pom.xml) wouldn't result in m2e picking up anything.

No, I mean m2e could implement a smarter project import: when "Import as Maven Project" finds `src/test/java/module-info.java` it could create the eclipse project configuration as shown in my example without any user interaction (or controlled by some checkbox in the wizard).

**Comment by danthe1st** [comment]
I just want to inform you about a [discussion on the Apache maven-dev mailing list](https://lists.apache.org/thread/zbzn5z55go6hqzyrnf2b3dnjvpl1s3rq). It might be a good idea that someone from JDT/m2e joins this discussion to come to a common/good solution.

I have also forwarded that E-Mail to the [eclipse-dev mailing list](https://www.eclipse.org/lists/eclipse-dev/msg12341.html), this thread could be useful as well.  


>>> During Devoxx Nicolai Parlog shared his concerns about how different tools
solve the modulepath/classpath in their own way (And he has a compliment:
Maven does it correct).

We were able to talk with the Jetbrains developers that are responsible for
Java/Maven integration and we're setting up a small group around this topic.
Maybe other tools might join too.

I'll be joining on behalf of Maven, but is there somebody else who would
like to join?

**Comment by nlisker** [comment]
I see that Maven is being discussed, but Gradle is neglected. Here is a thread regarding Gradle's side: https://discuss.gradle.org/t/testfixtures-dont-work-with-implementation-in-eclipse-only/47275/9?u=omega09.

Gradle uses multiple paths per *source set*, which are more or less equivalent for to Eclipse's source folders. The main source set is the source folders `src/main/java`, `src/main/resources`...; the test source set is `src/test/java`, `src/test/resources`...
But Gradle allows more than these, for example, to separate tests into unit tests, integrations test, functional tests etc. These might require different configurations as their boundaries are different. Gradle has [`testFixtures`](https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures), which is a source set for sharing tests. These types of tests would be under
`src/integrationTest/java`
`src/textFixtures/java`
...

Eclipse's idiom would be to create a different project altogether, but the test fixtures source set is usually logically part of a project with main sources.

*I think* that Gradle would have you treat each source set with its own module-info (or no module-info), not just main and test, like I believe Maven does. This might require someone who is more familiar with Gradle to elaborate on.

**Comment by nlisker** [comment]
Additional info on Gradle whitebox and blackbox testing and some notes on Eclipse: https://docs.gradle.org/current/userguide/java_testing.html#sec:java_testing_modular.

**Comment by hohwille** [comment]
Just for the record: A lot of people discussing such features do not really actively use JPMS.
You should also be aware that `module-info.java` allows direct service-loader integration (so aspects that otherwise OSGi or CDI frameworks mange).
Further, Java meanwhile supports access to the `ModuleLayer` for auto-discovery of components.
So in my JPMS project it is the new normal, that within `src/test/java` I place some "component" that needs to be exported or registered via module-info as otherwise it will not be found during the tests.
If tests are green in Maven but red in Eclipse this is no help.
And duplicating the structures and numbers of `pom.xml` files in my projects that are already complex and require quite some maintenance does not seem a real alternative...
I would still love if the different tools would have some common ground to build on.
I can fully understand that nobody is keen to start investing into a castle that is build on sand...

**Comment by stephan-herrmann** [comment]
I'd kindly ask every commenter to state whether they argue for the sake of
*  **black box** testing (tests are just another JPMS module, i.e., from JPMS p.o.v. tests are a regular client of the main module with no privileges), or
*  **white box** testing (tests are allowed to access internals of the main module that are not exposed to other modules).

If s.o. has already created a separated issue for either of these tests modes, please link it hear, so we can split the discussion.

To the best of my knowledge this is where these discussions have stalled:

### Black box:

Here we have several competing proposals:

* Obviously, Eclipse has always been able to support this mode using a **separate eclipse project**. This doesn't seem to be attractive for projects that are primarily declared in maven terms.
* In [this comment](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1465#issuecomment-1919835071)  I outlined steps that might allow to introduce a **separate compilation group** for compilation of tests with their own `module-info.java`  (search for "TEST_MODULE").  Significant, but feasible work. Yet no volunteer started working on this, right?
* In [this comment](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1465#issuecomment-2365280774) I suggested an approach using **nested eclipse projects**, where the main work would be in m2e to create this project structure from a given maven project. Plus the same for any other build tool like gradle.

### White box:

Perhaps [In-module testing with extra java command line options](https://sormuras.github.io/blog/2018-09-11-testing-in-the-modular-world#in-module-testing-with-extra-java-command-line-options) is still the most commonly agreed path, and the way to make this practical would be to store these options in a file `module-info.test` to be included into the commandline by the various tools' `@file` option. This approach needs no extension of Java/JPMS. 

In eclipse the current split in MAIN and TEST compilation groups already goes in this direction, and supporting a `module-info.test` file might boil down to replacing existing tweaks to the compiler invocation with arguments stored in that file.

I have also seen proposals to provide somewhat nicer source syntax from which that file could be generated. Once there is agreement across build tools / frameworks on such syntax, it shouldn't be too hard to support that in eclipse, too.

**Comment by danthe1st** [comment]
From what I can see, Maven does the following:
- If there's a `module-info.java` in `src/main/java` but not in `src/test/java`, module patching is used (from what I see, the surefire plugin adds test dependencies on the classpath, the module is patched with the test classes and it adds some arguments making (test) code accessible).
- If there is a `module-info.java` in both `src/main/java` and `src/test/java`, it uses black-box testing with two modules.

In my opinion, the more important thing is black-box testing as I guess most people would have module descriptors in both `src/main/java` and `src/test/java`. While I think having two projects would be fine (I guess having a distinct compilation group may be better from a user-experience perspective in the long run(?)), that shouldn't be too invasive in terms of what is shown to the user in my opinion (e.g. the package explorer should probably show the "normal" structure with the test sources being nested).

White-box testing would be useful as well but in my opinion, having black box testing is probably more important.

But I guess we should maybe ask people working on m2e for their opinions on what they think they need from JDT to properly support that? e.g. @laeubi

**Comment by stephan-herrmann** [comment]
@danthe1st  you spoke well, but we seem to have in imbalance between wishes and offers :)

For the proposed way of modular compilation groups within a project, someone would have to step forward to really get their hands dirty in enhancing JDT.  I might add a note that #4268 (which I know you are aware of) is still waiting for someone to implement/fix a quite similar feature in the context of multi-release projects. Whether one subsumes the other I cannot tell at this point, but at least both activities should be coordinated, to avoid any conflicts and/or unneeded code duplication.

I made the proposal using nested projects essentially for the case that we have to make do with JDT's current capabilities. It should be possible for m2e to implement an import wizard producing this structure without any enhancements in JDT, I believe. But maybe before doing that some people should experiment with that project structure to find out if this would work in real life, maybe with modifications.

**Comment by stephan-herrmann** [comment]
To unblock the case of white box testing, I filed #5079 which should be actionable right now.

**Comment by laeubi** [comment]
m2e would only be affected if there is some project settings to set, generally this looks like something that (at best) is discovered by jdt automatically as `src/main/java` and `src/test/java` are already distinguishable by the existing `test=true` classpath attribute.

The main "blocker" for me is that I'm not really using modules on my daily work and none of my customers do either as OSGi offers all I need regarding modularization already. Therefore I can not really judge what is the best approach here for implementation or what are the impacts.

In general during my Multi-Release development I found it quite useful to first start with an example project (or better a test-case), then see what error is produced and set a break point there. Then one can look through the call-stack and surrounding code if there is an opportunity to decide between the options and start with a first implementation idea to see what else then goes possibly wrong or to see the blast-radius of the change.

**Comment by danthe1st** [comment]
Since you mentioned an example project/test case: Here is a minimal example project with Maven and JPMS: https://github.com/danthe1st/m2e-jpms-repro
Maven detects the `module.info.java` files in both `src/main/java` and `src/test/java` to make sure these are their own modules. With Eclipse, both are part of the same module and it fails because:
- The `module-info.java` is present twice.
- It only considers the one from `src/main/java` hence imports to JUnit or similar are not accessible.
To fix that, m2e would need to create a nested second project (possibly with a bit of support on JDT-UIs side to not show it as distinct projects) if there is a `module-info.java` for both main and test source folders. See [this comment] from stephen.

I also created another branch [here](https://github.com/danthe1st/m2e-jpms-repro/tree/no-test-module-descriptor) with only a `module-info.java` in `src/main/java` but not `src/test/java`. This project works with Maven (`mvn verify`) as well but Eclipse doesn't do anything to make the tests available.
Maven (surefire) uses the following arguments to run the tests:
```
--module-path
"PROJECT/target/test-classes:PROJECT/target/classes:M2_REPOSITORY/org/junit/jupiter/junit-jupiter-api/6.0.3/junit-jupiter-api-6.0.3.jar:M2_REPOSITORY/org/opentest4j/opentest4j/1.3.0/opentest4j-1.3.0.jar:M2_REPOSITORY/org/junit/platform/junit-platform-commons/6.0.3/junit-platform-commons-6.0.3.jar:M2_REPOSITORYy/org/apiguardian/apiguardian-api/1.1.2/apiguardian-api-1.1.2.jar:M2_REPOSITORY/org/jspecify/jspecify/1.0.0/jspecify-1.0.0.jar"
--class-path
"M2_REPOSITORY/org/apache/maven/surefire/surefire-booter/3.5.2/surefire-booter-3.5.2.jar:/M2_REPOSITORY/org/apache/maven/surefire/surefire-api/3.5.2/surefire-api-3.5.2.jar:M2_REPOSITORY/org/apache/maven/surefire/surefire-extensions-spi/3.5.2/surefire-extensions-spi-3.5.2.jar:M2_REPOSITORY/org/apache/maven/surefire/surefire-shared-utils/3.5.2/surefire-shared-utils-3.5.2.jar:M2_REPOSITORY/org/apache/maven/surefire/surefire-logger-api/3.5.2/surefire-logger-api-3.5.2.jar:M2_REPOSITORY/org/apache/maven/surefire/surefire-junit-platform/3.5.2/surefire-junit-platform-3.5.2.jar:M2_REPOSITORY/org/apache/maven/surefire/common-java5/3.5.2/common-java5-3.5.2.jar:M2_REPOSITORY/org/junit/platform/junit-platform-engine/6.0.3/junit-platform-engine-6.0.3.jar:M2_REPOSITORY/org/junit/platform/junit-platform-launcher/6.0.3/junit-platform-launcher-6.0.3.jar:M2_REPOSITORY/org/junit/jupiter/junit-jupiter-engine/6.0.3/junit-jupiter-engine-6.0.3.jar"
--add-modules
ALL-MODULE-PATH
--add-opens
org.junit.platform.commons/org.junit.platform.commons.util=ALL-UNNAMED
--add-opens
org.junit.platform.commons/org.junit.platform.commons.logging=ALL-UNNAMED
org.apache.maven.surefire.booter.ForkedBooter
```

And for compiling it uses the following:
```
-d PROJECT/target/test-classes --module-path PROJECT/target/test-classes:PROJECT/target/classes:M2_REPOSITORY/org/junit/jupiter/junit-jupiter-api/6.0.3/junit-jupiter-api-6.0.3.jar:M2_REPOSITORY/org/opentest4j/opentest4j/1.3.0/opentest4j-1.3.0.jar:M2_REPOSITORY/org/junit/platform/junit-platform-commons/6.0.3/junit-platform-commons-6.0.3.jar:M2_REPOSITORY/org/apiguardian/apiguardian-api/1.1.2/apiguardian-api-1.1.2.jar:M2_REPOSITORY/org/jspecify/jspecify/1.0.0/jspecify-1.0.0.jar: -sourcepath PROJECT/src/test/java: -s PROJECT/target/generated-test-sources/test-annotations -g --release 21 -encoding UTF-8 -XDuseUnsharedTable=true
```

The [link stephen shared](https://sormuras.github.io/blog/2018-09-11-testing-in-the-modular-world#in-module-testing-with-extra-java-command-line-options) may be useful for this.

With Eclipse, both versions fail to compile.

**Comment by danthe1st** [comment]
I also created [another branch](https://github.com/danthe1st/m2e-jpms-repro/tree/working-no-use-module-path) without a `module-info.java` in `src/test/java` but with the `<useModulePath>false</useModulePath>` option in the surefire plugin `<configuration>`. With this, it compiles and the test passes in Eclipse (though `mvn verify` fails as my test expects a module). I think whatever m2e does with `<useModulePath>false</useModulePath>` might be the correct option if the is a `module-info.java` in `src/main/java` but not `src/test/java`.
I think Eclipse with `<useModulePath>false</useModulePath>` behaves like Maven with `<useModulePath>true</useModulePath>`.

Could you verify that? @laeubi 
Do you think it is reasonable to change m2e in a way that it works as it does currently when `<useModulePath>false</useModulePath>` is present even if it is?

**Comment by laeubi** [comment]
> With Eclipse, both are part of the same module and it fails because:
> The module-info.java is present twice.

This is something we have already solved for MR jar support so it would theoretically quite easy to get beyond that problem especially as these are two distinct compilation phases even.

> It only considers the one from src/main/java hence imports to JUnit or similar are not accessible.

Would the expectation be to combine the one from `main`+`test` or only use the one from `test` in case of test compilation? At least it sound doable with some classpath ordering tricks (we do similar for MR jar as well).

> Do you think it is reasonable to change m2e in a way that it works as it does currently when <useModulePath>false</useModulePath> is present even if it is?

This is a bit hard to tell, but in general I would like to avoid 'tricks' in m2e because they then only would work for maven/m2e but not if user want to use other toolings (including plain jdt).

**Comment by stephan-herrmann** [comment]
> Would the expectation be to combine the one from `main`+`test` or only use the one from `test` in case of test compilation? 

Since no standard defines how two module descriptions could be merged, we should not go down that road. 

Using only the one from test would be a way of white box testing, which I outsourced to #5079  (with less redundancy than specifying two complete module descriptions).

Ergo, the remaining issue here is:
* let a test source folder behave like any regular JPMS module. Using its own `module-info.java` it can
  * require the main module (which now is a black box)
  * add its own dependencies and service definitions

The two competing strategies to this goal are:
* maintain the concept that each eclipse project defines exactly one module
  * solve the main vs test issue using nested projects
  * JDT might support this out of the box
  * setting up eclipse projects in this way would be a task for a wizard outside JDT, e.g., m2e based on maven conventions
* enhance JDT such that the "test" compilation group could optionally define its own module
  * in contrast to MR compilation this would require that within the project both modules are visible in the same compilation, where the main module would be seen as "external" during compilation of "test" (perhaps by depending on the output path of main).
     * the existing option to depend on such a project _including test code_ would pose additional challenges, as this would mean we need to expose the presence of two modules in the project, which cannot easily be done without breaking changes
  * the existing strategy of implicit tweaks for compiling "test" should still remain as an option.

Long story short: neither merge both, nor hide main, but compile test with the main module as a dependency.

**Comment by danthe1st** [comment]
> This is a bit hard to tell, but in general I would like to avoid 'tricks' in m2e because they then only would work for maven/m2e but not if user want to use other toolings (including plain jdt).

The reason I asked for that is because Eclipse seems to behave correctly for the opposite option if it is set. If I set `<useModulePath>false</useModulePath>` and run a test, it uses CLI options like this:
```
java
-ea
--add-opens main/io.github.danthe1st.repro.m2ejpms.test=ALL-UNNAMED
--add-modules=ALL-MODULE-PATH
-Dfile.encoding=UTF-8
-Dstdout.encoding=UTF-8
-Dstderr.encoding=UTF-8
-p PROJECT/target/classes
-classpath M2_REPOSITORY/org/junit/jupiter/junit-jupiter-api/6.0.3/junit-jupiter-api-6.0.3.jar:M2_REPOSITORY/org/opentest4j/opentest4j/1.3.0/opentest4j-1.3.0.jar:M2_REPOSITORY/org/junit/platform/junit-platform-commons/6.0.3/junit-platform-commons-6.0.3.jar:M2_REPOSITORY/org/apiguardian/apiguardian-api/1.1.2/apiguardian-api-1.1.2.jar:M2_REPOSITORY/org/jspecify/jspecify/1.0.0/jspecify-1.0.0.jar:M2_REPOSITORY/org/junit/jupiter/junit-jupiter-engine/6.0.3/junit-jupiter-engine-6.0.3.jar:M2_REPOSITORY/org/junit/platform/junit-platform-engine/6.0.3/junit-platform-engine-6.0.3.jar:ECLIPSE_HOME/configuration/org.eclipse.osgi/2054/0/.cp:/ECLIPSE_HOME/configuration/org.eclipse.osgi/521/0/.cp:ECLIPSE_HOME/configuration/org.eclipse.osgi/1970/0/.cp
-XX:+ShowCodeDetailsInExceptionMessages
--patch-module main=PROJECT/target/test-classes
--add-reads main=ALL-UNNAMED org.eclipse.jdt.internal.junit.runner.RemoteTestRunner
-version 3
-port 37719
-testLoaderClass org.eclipse.jdt.internal.junit6.runner.JUnit6TestLoader
-loaderpluginname org.eclipse.jdt.junit6.runtime
-classNames io.github.danthe1st.repro.m2ejpms.test.TestClass
```
From what I can see, this looks more or less **like a correct configuration for white-box testing with `<useModulePath>true</useModulePath>`**

When I add modular non-test dependencies with `<useModulePath>false</useModulePath>` and run a test, it even runs the test with those dependencies on the modulepath (while Maven behaves like that with `<useModulePath>true</useModulePath>`):
```
-p PROJECT/target/classes:M2_REPOSITORY/org/slf4j/slf4j-api/2.0.17/slf4j-api-2.0.17.jar:M2_REPOSITORY/org/slf4j/slf4j-simple/2.0.17/slf4j-simple-2.0.17.jar
```

**Comment by laeubi** [comment]
From what I have made here for MR support (what would be ready to review/merge @stephan-herrmann if you find time):

- https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4534

I think we would need a method `IModuleDescription getModuleDescription(boolean test)` this can then (whatever convention/magic we decide) return the projects own descriptor, one derived automagically or one explicitly defined.

The main project source are already part of the classpath for the test compilation phase/group.

@danthe1st I'm not sure if this refers to m2e specifically or eclipse in general. In any case it might be usefull to derive a plain JDT test-case and submit it as a PR, then its easier to discuss on code if your desired effect is a bug or a feature or something incomplete implemented.

As on the maven channel there is some discussion around tooling support I asked if someone minds to chime in here as well.

**Comment by desruisseaux** [comment]
The following comment is about **white-box testing**. I ignore black-box testing here. All Maven examples that I saw in this thread were based on Maven 3. Things have changed a lot in Maven 4, and I argue that Maven 3 was actually a bad example to follow. I suggest the approach implemented in Maven 4 together with Maven-Compiler-Plugin 4.0.0-beta-4:

* **Never overwrite `module-info.java` for testing purpose**, unless there is no way to do otherwise. The only exception that I'm aware of is when we want to register a different set of service providers (the `provides` clause) or use a different set of services (the `uses` clause).
* Instead of overwriting `module-info.java`, use the `--add-modules`, `--add-reads`, `--add-exports` and `--add-opens` options. The absence of `--add-uses` and `--add-provides` options is the reason for the exception in above point, but I presume that even this exception could be resolved in a future Java version if Oracle decides to do so.
* Because above-cited options are tedious to write in a `pom.xml` file, make that easier for users. For example, the module name is repeated often if the options are declared manually. This redundancy can be avoided with a little bit of help from the build tool (Maven or other).

Maven 4 invented `module-info-patch.maven` (the `.maven` suffix is because this is specific to Maven at this time, but could be changed if there is an agreement to use something like that with more tools). The syntax is very similar to `module-info.java` with keywords almost identical to the `javac` options. This is described in the [Module-info patch page of compiler plugin](https://maven.apache.org/plugins/maven-compiler-plugin-4.x/module-info-patch.html).

Another reason for `module-info-patch.maven` which may not be relevant right now for Eclipse but may become relevant in the future, is that it integrates very well with module source hierarchy. Note that supporting module source hierarchy also leads naturally to a better support of multi-release projects (both aspects can be handled in same time). Description on the [Modular projects page of compiler plugin](https://maven.apache.org/plugins/maven-compiler-plugin-4.x/modules.html).

Maven Compiler Plugin 4.0.0-beta-4 fully supports the above and is deployed on Maven Central since January. Some changes are required in the JAR plugin - a pull request is ready. Some more changes are required by the Surefire plugin for using the options that the compiler plugin prepared for it.

This approach is in the same spirit as the [Christian Stein's blog](https://sormuras.github.io/blog/2018-09-11-testing-in-the-modular-world#in-module-testing-with-extra-java-command-line-options) mentioned in other comments of this thread. The way to do that is different, but the core principle is the same.
